keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,305 @@
+
+### Goals
+
+- When CSI Driver is installed on the node, same predicate will be used to handle Volume limit counting for in-tree as well as CSI Volumes. Same limit will be used when user is using both CSI and in-tree volumes on the node.
+
+- Existing predicates for in-tree volumes `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount`(now deprecated) will be removed when in-tree to CSI migration is enabled for respective volume plugin.
+  - When both deprecated in-tree predicate and CSI predicate are enabled, only `MaxCSIVolumeCountPred` does useful work and the other is NOOP to save CPU. This requires CSI Driver to be installed on the node.
+
+- Scheduler does not put pods that require CSI volumes to nodes that don't have the CSI driver installed.
+
+- Scheduler does not increase its CPU consumption. Any regression must be approved by sig-scheduling.
+  - Scheduler benchmark must be extended to schedule pods with volumes as part of this KEP.
+
+### Non-Goals
+
+- Heterogenous clusters, i.e. clusters where access to storage is limited only to some nodes. Existing `PV.spec.nodeAffinity` handling, not modified by this KEP, will filter out nodes that don't have access to the storage, so predicates changed in this KEP don't need to worry about storage topology and can be simpler.
+
+- Multiple plugins sharing the same volume limits. We expect that every CSI driver will have its own limits, not shared with other CSI drivers. In this KEP we support only in-tree volume plugins sharing its limits with one hard-coded CSI driver each.
+
+- Multiple "units" per single volume. Each volume used on a node takes exactly 1 unit from `allocatable.volumes`, regardless of the volume size, its replica count, number of connections to remote servers or other underlying resources needed to use the volume. For example, multipath iSCSI volume with three paths (and thus three iSCSI connections to three different servers) still takes 1 unit from `CSINode.spec.drivers[xyz].allocatable.volumes`.
+
+- Maximum capacity per node. Some cloud environments limit both number of attached volumes (covered in this KEP) and total capacity of attached volumes (not covered in this KEP). For example, this KEP will ensure that scheduler puts max. 128 GCE PD volumes to a [typical GCE node](https://cloud.google.com/compute/docs/machine-types#predefined_machine_types), but it won't ensure that the total capacity of the volumes is less than 64 TB.
+
+- Volume limits does not yet integrate with cluster autoscaler if all nodes in the cluster are running at maximum volume limits.
+
+## Proposal
+
+* Track volume limits for CSI driver in `CSINode` objects instead of `Node`. Limit in `CSINode` is used in-stead of limit coming from `Node` object whenever available for same in-tree volume type.
```

Comment:
typo: instead

Reply:
Add a note here that that means in the scheduler, we will always try to translate in-tree driver name to CSI driver name, regardless of if migration is on or off.

Reply:
fixed